src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs

```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+                        ErrorRecord errorRecord = new ErrorRecord(
+                            new Exception(errorText),
+                            "SOURCE_CODE_ERROR",
+                            ErrorCategory.InvalidData,
+                            diagnisticRecord);
+
+                        WriteError(errorRecord);
+                    }
+                }
+
+                if (IsError)
+                {
+                    ErrorRecord errorRecord = new ErrorRecord(
+                        new InvalidOperationException(AddTypeStrings.CompilerErrors),
+                        "COMPILER_ERRORS",
+                        ErrorCategory.InvalidData,
+                        null);
+                    ThrowTerminatingError(errorRecord);
+                }
             }
         }
 
-        private AddTypeCompilerError[] GetErrors(ImmutableArray<Diagnostic> diagnostics)
+        private string BuildErrorMessage(Diagnostic diagnisticRecord)
         {
-            return diagnostics.Where(d => d.Severity == DiagnosticSeverity.Error || d.Severity == DiagnosticSeverity.Warning)
-                .Select(d => new AddTypeCompilerError
+            var location = diagnisticRecord.Location;
+
+            if (location.SourceTree == null)
+            {
+                // For some error types (linker?) we don't have related source code.
+                return diagnisticRecord.ToString();
+            }
+            else
+            {
+                var text = location.SourceTree.GetText();
+                var textLines = text.Lines;
+
+                var lineSpan = location.GetLineSpan(); // FileLinePositionSpan type.
+                var errorLineNumber = lineSpan.StartLinePosition.Line;
+                var errorPosition = lineSpan.StartLinePosition.Character;
+
+                // This is typical Roslyn diagnostic message which contains
+                // a message number, a source context and an error position.
+                var diagnisticMessage = diagnisticRecord.ToString();
+                var errorLineString = textLines[errorLineNumber].ToString();
+
+                StringBuilder sb = new StringBuilder(diagnisticMessage.Length + errorLineString.Length * 2 + 4);
+
+                sb.AppendLine(diagnisticMessage);
+                sb.AppendLine(errorLineString);
+
+                for (var i = 0 ; i < errorLineString.Length ; i++)
                 {
-                    ErrorText = d.GetMessage(),
-                    FileName = null,
-                    Line = d.Location.GetMappedLineSpan().StartLinePosition.Line + 1, // Convert 0-based to 1-based
-                    IsWarning = !d.IsWarningAsError && d.Severity == DiagnosticSeverity.Warning,
-                    Column = d.Location.GetMappedLineSpan().StartLinePosition.Character + 1, // Convert 0-based to 1-based
-                    ErrorNumber = d.Id
-                }).ToArray();
+                    if (errorLineString[i] != ' ' && errorLineString[i] != '\t')
```

Comment:
Maybe change to `if (!char.IsWhiteSpace(errorLineString[i]))`? It's more readable.

Reply:
I thought about it but there's a lot of other characters
https://msdn.microsoft.com/en-us/library/t809ektx%28v=vs.110%29.aspx
Is it ok?

Reply:
As long as they are whitespace characters, I don't see a difference. Do you?

Reply:
I fixed this but ...
This come from original code and I'm afraid the code isn't exactly correct because we're replacing one tab with a single space. Perhaps it was necessary for CodeDom. I hope that Roslyn don't use tabs in error messages. Although these whitespaces can come from the code that it compiles. Therefore we have to clean the error message too. Thoughts?

Reply:
Fixed.